src/tactic/ring_exp.lean

```diff
@@ -0,0 +1,1430 @@
+  - In the tactic, a normalized expression `ps : ex` lives in the meta-world,
+    but the normalization proofs live in the real world.
+    Thus, we cannot directly say `ps.orig = ps.pretty` anywhere,
+    but we have to carefully construct the proof when we compute `ps`.
+    This was a major source of bugs in development!
+  - For `pow`, the exponent must be a natural number, while the base can be any semiring `α`.
+    We swap out operations for the base ring `α` with those for the exponent ring `ℕ`
+    as soon as we deal with exponents.
+    This is accomplished by the `in_exponent` function and is relatively painless since
+    we work in a `reader` monad.
+  - The normalized form of an expression is the one that is useful for the tactic,
+    but not as nice to read. To remedy this, the user-facing normalization calls `ex.simp`.
+
+  ## Caveats and future work
+
+  Subtraction cancels out identical terms, but division does not.
+  That is: `a - a = 0 := by ring_exp` solves the goal,
+  but `a / a := 1 by ring_exp` doesn't.
+  Note that `0 / 0` is generally defined to be `0`,
+  so division cancelling out is not true in general.
+
+  Multiplication of powers can be simplified a little bit further:
+  `2 ^ n * 2 ^ n = 4 ^ n := by ring_exp` could be implemented
+  in a similar way that `2 * a + 2 * a = 4 * a := by ring_exp` already works.
+  This feature wasn't needed yet, so it's not implemented yet.
+
+  ## Tags
+
+  ring, semiring, exponent, power
+-/
+
+-- The base ring `α` will have a universe level `u`.
+-- We do not introduce `α` as a variable yet,
+-- in order to make it explicit or implicit as required.
+universes u
+
+namespace tactic.ring_exp
+open nat
+
+/--
+  The `atom` structure is used to represent atomic expressions:
+  those which `ring_exp` cannot parse any further.
+
+  For instance, `a + (a % b)` has `a` and `(a % b)` as atoms.
+  The `ring_exp_eq` tactic does not normalize the subexpressions in atoms,
+  but `ring_exp` does if `ring_exp_eq` was not sufficient.
+
+  Atoms in fact represent equivalence classes of expressions,
+  modulo definitional equality.
+  The field `index : ℕ` should be a unique number for each class,
+  while `value : expr` contains a representative of this class.
+  The function `resolve_atom` determines the appropriate atom
+  for a given expression.
+-/
+meta structure atom : Type := (value : expr) (index : ℕ)
+namespace atom
+/--
+  The `eq` operation on `atom`s works modulo definitional equality,
+  ignoring their `value`s.
+  The invariants on `atom` ensure indices are unique per value.
+  Thus, `eq` indicates equality as long as the `atom`s come from the same context.
+-/
+meta def eq (a b : atom) : bool := a.index = b.index
+/--
+  We order `atom`s on the order of appearance in the main expression.
+-/
+meta def lt (a b : atom) : bool := a.index < b.index
+meta instance : has_repr atom := ⟨λ x, "(atom " ++ repr x.2 ++ ")"⟩
+end atom
+
+section expression
+/-!
+  ## `expression` section
```

Comment:
```suggestion
  ### `expression` section
```
Changing my mind from before. Since we use h2 headers in the header docs, using h3 internally seems better.